modules/exploits/linux/local/cve_2022_0995_watch_queue.rb

```diff
@@ -0,0 +1,162 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = GreatRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Post::Linux::Priv
+  include Msf::Post::Linux::System
+  include Msf::Post::Linux::Compile
+  include Msf::Post::Linux::Kernel
+  include Msf::Post::File
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Watch Queue Out of Bounds Write',
+        'Description' => %q{
+          This module exploits a vulnerability in the Linux Kernel's watch_queue event notification system. It relies on a heap
+          out-of-bounds write in kernel memory. The exploit fails occasionally, but it does not crash the target and can be run
+          multiple times until privilege escalation is achieved.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'bonfee',     # PoC
+          'bwatters-r7' # Aka @tychos_moose, Metasploit Module
+        ],
+        'DisclosureDate' => '2022-03-14',
+        'Platform' => [ 'linux' ],
+        'Arch' => [ ARCH_X64 ],
+        'SessionTypes' => [ 'shell', 'meterpreter' ],
+        'Privileged' => true,
+        'References' => [
+          [ 'CVE', '2022-0995' ],
+          [ 'URL', 'https://github.com/Bonfee/CVE-2022-0995' ],
+          [ 'URL', 'https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=93ce93587d36493f2f86921fa79921b3cba63fbb' ],
+          [ 'URL', 'https://nvd.nist.gov/vuln/detail/CVE-2022-0995' ]
+        ],
+        'Targets' => [
+          [ 'Ubuntu Linux 5.13.0-37', {} ],
+        ],
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ CRASH_OS_RESTARTS ], # In practice haven't seen this happen but there is a small potential for a kernel panic here.
```

Comment:
Also, failure due to corrupted IPC queue (unlikely, but eventually inevitable following multiple exploitation attempts) will result in system instability, including potential for system freeze.

```suggestion
          'Stability' => [ CRASH_OS_DOWN ],
```

`CRASH_OS_DOWN` is consistent with other Linux kernel memory corruption exploits.

```
# grep -rn CRASH_OS_ modules/exploits/linux/local/
modules/exploits/linux/local/bpf_sign_extension_priv_esc.rb:93:            'Stability' => [ CRASH_OS_DOWN ]
modules/exploits/linux/local/sock_sendpage.rb:67:          'Stability'   => [ CRASH_OS_DOWN ],
modules/exploits/linux/local/cve_2021_3490_ebpf_alu32_bounds_check_lpe.rb:65:          'Stability' => [ CRASH_OS_DOWN ],
modules/exploits/linux/local/dirtycow_priv_esc.rb:76:            'Stability' => [ CRASH_OS_DOWN ]
modules/exploits/linux/local/rds_atomic_free_op_null_pointer_deref_priv_esc.rb:75:          'Stability'   => [ CRASH_OS_DOWN ],
modules/exploits/linux/local/af_packet_packet_set_ring_priv_esc.rb:73:          'Stability'   => [ CRASH_OS_DOWN ],
modules/exploits/linux/local/af_packet_chocobo_root_priv_esc.rb:74:          'Stability'   => [ CRASH_OS_DOWN ]
modules/exploits/linux/local/ufo_privilege_escalation.rb:72:          'Stability'   => [ CRASH_OS_DOWN ],
modules/exploits/linux/local/netfilter_priv_esc_ipv4.rb:59:            'Stability'   => [ CRASH_OS_DOWN ],
```

Except this one, which really should probably use `CRASH_OS_DOWN` too.

```
# grep -rn Stability modules/exploits/linux/local/  | fgrep -v CRAS
modules/exploits/linux/local/netfilter_xtables_heap_oob_write_priv_esc.rb:45:          'Stability' => [ OS_RESOURCE_LOSS ], # After too many failed attempts, the system needs to be restarted.
modules/exploits/linux/local/cve_2021_3493_overlayfs.rb:49:          'Stability' => [ ],
```

Reply:
Makes sense to me.  I've updated my local branch and I'll wait a bit to see if there's anything else that needs changing before I push it again.

Reply:
Sounds good to me @bcoles thanks for pointing that out 👍